### ISSUE TITLE
Clipped highlights look posterized

### ISSUE THREAD

**Issue by TooWaBoo** [Beep6581/RawTherapee#4479]
Images taken with my pocket cam **Nikon P330** and clipped highlights look posterized and washed out. In the past it looked fine, but with the latest version of RT it doesn't. I can't tell when this issue appeared the first time.

Link to RAW and pp3
https://filebin.net/qxlzm3k1w6g087je

![grafik](https://user-images.githubusercontent.com/13800920/38162880-d4648a48-34e9-11e8-9401-4f46fdc63c46.png)

```
Version: 5.4-122-gf9396f2a
Branch: dev
Commit: f9396f2a
Commit date: 2018-03-29
Compiler: gcc 7.3.0
Processor: undefined
System: Windows
Bit depth: 64 bits
Gtkmm: V3.22.0
Lensfun: V0.3.2.0
Build type: release
Build flags: -mwin32 -m64 -mthreads -msse2 -std=c++11 -march=native -Werror=unused-label -fopenmp -Werror=unknown-pragmas -Wall -Wno-unused-result -Wno-deprecated-declarations -Wno-aggressive-loop-optimizations -DNDEBUG -O3 -ftree-vectorize
Link flags: -m64 -mthreads -static-libgcc -fno-use-linker-plugin -march=native -s -O3
OpenMP support: ON
MMAP support: ON
```

**Comment by Floessie** [comment]
Could be related to #4457...

**Comment by Beep6581** [comment]
Cannot reproduce in `5.4-122-g`f9396f2a
![imgur-2018_03_31-14 57 50](https://user-images.githubusercontent.com/5844619/38163407-f853905c-34f3-11e8-8356-552e73e349fb.png)

**Comment by Beep6581** [comment]
@Floessie I thought so too, but the PP3 uses `method=blend`, and changing it to "color propagation" works well.

**Comment by Floessie** [comment]
@Beep6581 I see. :+1: Sorry for the noise.

**Comment by TooWaBoo** [comment]
I get this when changing to "color propagation"

![grafik](https://user-images.githubusercontent.com/13800920/38163907-e31d6034-34fb-11e8-8fa6-98d6092cbcb1.png)

**Comment by Beep6581** [comment]
@TooWaBoo can you try to eliminate unrelated tools? i.e. start from "Neutral" and find which tool causes that. I would suggest you delete or rename your options file as well, to make sure we're using the same settings.

**Comment by TooWaBoo** [comment]
@Beep6581 
I've followed your instuctions.
I think I found the culprit. It's the Profiled Lens correction -> Vignetting correction
I've uploaded the .lcp-file to the link above too.

How to reproduce:
1. Apply Neutral
2. Check Highlight reconstruction
3. Select P330...lcp
4. Uncheck Distortion correction for better preview
5. Check Vignetting correction
6. The sky becomes posterized

There's another issue. If you then check radio button "None", Vignetting correction is still active.

**Comment by Beep6581** [comment]
Confirmed both issues.
There is a third issue - using neutral, there are strong magenta highlights.

**Comment by Beep6581** [comment]
The posterized highlight transition is unrelated to LCP.
Simply apply neutral and set white-point correction to 2.
![screenshot_20180331_170745](https://user-images.githubusercontent.com/5844619/38164506-0ca0d24c-3506-11e8-8f6e-8f7d19a6a37d.png)

**Comment by TooWaBoo** [comment]
The strong magenta highlights will disappear by acivating highlight reconstruction.

**Comment by cuniek** [comment]
Isn't that the result of unbounded processing?
EDIT: Scratch that, this is wrong white point (as @Beep6581 wrote). I made the comment because I am evaluating unbounded processing which introduced pink highlights in cameras that do not suffer from wrong whitepoint.

**Comment by Beep6581** [comment]
I measured the white/black levels, tried this and that, and the white level number which eliminates magenta is nowhere near what I measured, even after scaling the measured value to 14-bit... @iliasg help, could you explain how to measure this file correctly?

```javascript
    { // Quality X
        "make_model": "NIKON COOLPIX P330",
        "dcraw_matrix": [ 10321,-3920,-931,-2750,11146,1824,-442,1545,5539 ], // Adobe DNG Converter 10.2
        "ranges": { "white": 12200 }
    },

```

**Comment by agriggio** [comment]
could be a side effect of unbounded processing, if you don't enable highlight reconstruction. RT used to clip highlights, now it doesn't unless told so (ie by enabling highlight recovery). I see a couple of options, but I'm open to suggestions:

1. restore the old behaviour
2. introduce a new highlight reconstruction mode "clip"
3. live with this, as there are workarounds

(I'm fine with 3 personally)

**Comment by TooWaBoo** [comment]
I vote for 1. restore the old behaviour
I've tried to get the same results with the new behaviour, but I could'nt get it identical. (I've compared with RT5.3)

**Comment by agriggio** [comment]
@TooWaBoo what do you mean compared with 5.3? what I described above happened after 5.4, so the reference point should be 5.4 -- otherwise the issue is different... can you confirm that with 5.4 everything works as you expected?

**Comment by TooWaBoo** [comment]
@agriggio 
I'll test with 5.4 tomorrow.

**Comment by Beep6581** [comment]
RT 5.3:
![imgur-2018_04_01-01 15 13](https://user-images.githubusercontent.com/5844619/38168366-2a797a78-354a-11e8-8489-846ab1a6e90e.png)

RT dev 69c38b45 (before the unbounded merge):
![imgur-2018_04_01-01 16 19](https://user-images.githubusercontent.com/5844619/38168376-5f2bce24-354a-11e8-9698-537fa2de50c8.png)

RT dev ed21c4ee
![imgur-2018_04_01-01 18 17](https://user-images.githubusercontent.com/5844619/38168383-9753ad9e-354a-11e8-885c-031278a3898d.png)

**Comment by iliasg** [comment]
@Beep6581 
> help, could you explain how to measure this file correctly?

Where did this WL = 12200 came from !!

On the heavylly overexposed areas the clipping is at 4094 (just one down from the 12bit top 4095 !!) .. I would put it a bit lower (say 4080) to be on the safe side

@agriggio 
But with RT 5.4-84-gc4933e36 we have a problem with DNG where adobe converts the NRW to 16bit DNG BL=3200, WL=65000 (not sure if this is already OK after Alberto's related to DNG WL commit) 

BTW I see the white level range is restricted to 0.10-16.0 .. and this is not enough to correct the WL manually in this case (0.062 needed ..) .. I think this slider should better be in log2( ) scale (see code for pixelshift Iso adaption)

**Comment by TooWaBoo** [comment]
@agriggio 
I've tested with 5.4 and everything is fine.

**Comment by agriggio** [comment]
so is the pink building the desired behaviour?
how about a checkbox in the raw tab that lets the user decide whether to clip or not?

**Comment by TooWaBoo** [comment]
@agriggio 
As long as I get the old behaviour back, you can do what ever you want. 😁
Now serious: I don't understand the benefit of the new behavior.

**Comment by TooWaBoo** [comment]
There is an issue in the shadows too.
dev (new)
![grafik](https://user-images.githubusercontent.com/13800920/38171078-5dfaf9ec-3592-11e8-97cc-c227d25ed8d5.png)

5.4 (old)
![grafik](https://user-images.githubusercontent.com/13800920/38171088-b723d82c-3592-11e8-89f3-a2a8862bed49.png)

**Comment by agriggio** [comment]
@TooWaBoo https://filebin.net/qxlzm3k1w6g087je/DSCN0760-2.jpg
For the shadows part, could you please attach a RAW file and a pp3, so that I can have a look? Thanks!

**Comment by TooWaBoo** [comment]
Here it is: https://filebin.net/08j6p7pn1btgx5ms

**Comment by Beep6581** [comment]
> On the heavylly overexposed areas the clipping is at 4094 (just one down from the 12bit top 4095 !!) .. I would put it a bit lower (say 4080) to be on the safe side

Exactly, but when I tried 4094 and 3894, it was as if RT ignored it. Then I tried scaling that value to a 14-bit range, RT didn't ignore it, but it still showed magenta.

**Comment by cuniek** [comment]
@TooWaBoo Change input profile from "Camera standard" to any icc, dcp, or "No profile". The problem is gone.

**Comment by TooWaBoo** [comment]
@cuniek 
But this wouldn't really help. I'm using the P330...dcp from Adobe and the problem still exists with latest dev.

**Comment by cuniek** [comment]
@TooWaBoo
Here is a set of screenshots. I have tested different profiles (Nikon d750, Sony 580, Minolta 7D, even AdobeRGB), and 90% of them give smooth shadows. There are few that give wrong shadows, so... This is either color profile related, or black point in RT is wrong, but some color profiles are correcting that.

Notice, that Raw Black Points are fixing it too.

![2018-04-01_133652](https://user-images.githubusercontent.com/26029842/38172721-d51d2bfc-35b1-11e8-8ac3-f1e67dc602d5.jpg)

![2018-04-01_133124](https://user-images.githubusercontent.com/26029842/38172724-da457aee-35b1-11e8-83d7-7bb45e6c2734.jpg)

![2018-04-01_133101](https://user-images.githubusercontent.com/26029842/38172733-ee78facc-35b1-11e8-98fd-b357629b6be3.jpg)

**Comment by agriggio** [comment]
btw, I don't want to advocate the change at any cost (in fact I think a checkbox in the raw tab is the way to go), but similar posterization can be obtained using 5.4 on different shots, at least as far as I remember -- I'll dig deeper once I get back to my machine

**Comment by Beep6581** [comment]
Three different issues are being discussed:

1. Problem 1: incorrect white levels. Setting `"white": 4080` works fine for the NRW but not for DNG in `dev` (`5.4-127-g`ed21c4ee at the time of writing).
2. Problem 2: LCP vignetting correction is applied even when profiled lens correction is set to "None" - I opened issue #4480
3. Problem 3: highlight transition to clipping is not smooth.

Let's please limit the discussion here to problem 3, as that is the title of this issue, and open a new issue for problem 1.

**Comment by cuniek** [comment]
@Beep6581 
About problem 3 - I guess I have found the cause 

line 1112

https://github.com/Beep6581/RawTherapee/commit/15813abbedb6079a0f3ec627675dcf8385939de1#commitcomment-28366769

**Comment by agriggio** [comment]
@cuniek according to @TooWaBoo the problem occurs with "blend", not "colour propagation"...

**Comment by cuniek** [comment]
Last post for today, I promise:

@agriggio You are (somehow) right, I made some comparisons and ALL HR methods are giving posterized highlights after unbounded was commited. Still the issue I mentioned adds even more posterization in case of colour propagation.

Happy Easter!

**Comment by agriggio** [comment]
@cuniek no problem, please feel free to add as many comments you feel appropriate! 
Regarding the posterization: the "trick" is to use highlight compression. Do you see posterization in the file I posted above? (https://filebin.net/qxlzm3k1w6g087je/DSCN0760-2.jpg). The pp3 is here:
https://filebin.net/qxlzm3k1w6g087je/DSCN0760-2.jpg.out.pp3
It looks fine from here (but again, I'm on a crappy screen).

**Comment by Beep6581** [comment]
There is a test image for examining transitions from unclipped to all-clipped:
http://rawtherapee.com/shared/test_images/overexposed_sky.pef

Neutral

* 5.3
![screenshot_20180401_161227](https://user-images.githubusercontent.com/5844619/38174094-5e19d52a-35c8-11e8-873f-ef110d33fd07.png)
* 5.4
![screenshot_20180401_161304](https://user-images.githubusercontent.com/5844619/38174097-68939626-35c8-11e8-9766-07f024b006e8.png)
* 69c38b45 (the commit before `unbounded-processing` was merged into `dev`) 
![screenshot_20180401_161332](https://user-images.githubusercontent.com/5844619/38174099-6ef562d8-35c8-11e8-89dd-eee5a579012f.png)
* Latest `dev` ed21c4ee
![screenshot_20180401_163201](https://user-images.githubusercontent.com/5844619/38174210-3977def4-35ca-11e8-9dcb-6988f239c356.png)

**Comment by Beep6581** [comment]
Neutral + color propagation

* 5.3
![screenshot_20180401_163736](https://user-images.githubusercontent.com/5844619/38174288-4261fc74-35cb-11e8-8e2d-e30ebc630742.png)
* 5.4
![screenshot_20180401_163805](https://user-images.githubusercontent.com/5844619/38174293-4634f766-35cb-11e8-8bfd-0b81498111ef.png)
* 69c38b45 (the commit before `unbounded-processing` was merged into `dev`) 
![screenshot_20180401_163831](https://user-images.githubusercontent.com/5844619/38174294-4933859a-35cb-11e8-875d-aad9f2ed8122.png)
* Latest `dev` ed21c4ee
![screenshot_20180401_163901](https://user-images.githubusercontent.com/5844619/38174295-4c2c5fa6-35cb-11e8-9721-da136a219b0a.png)

**Comment by agriggio** [comment]
I will add the checkboxes to let the user decide whether to clip (default old behaviour)

**Comment by agriggio** [comment]
No need for extra checkboxes. Here's the patch that solves the problem for the highlights:

```diff
diff --git a/rtengine/improcfun.cc b/rtengine/improcfun.cc
--- a/rtengine/improcfun.cc
+++ b/rtengine/improcfun.cc
@@ -3764,13 +3764,12 @@
                         float g = std::max(gtemp[ti * TS + tj], 0.f);
                         float b = std::max(btemp[ti * TS + tj], 0.f);
 
-                        if (r > 65535 || g > 65535 || b > 65535) {
+                        if (max(r, g, b) > MAXVALF && min(r, g, b) < MAXVALF) {
                             filmlike_clip (&r, &g, &b);
+                            rtemp[ti * TS + tj] = r;
+                            gtemp[ti * TS + tj] = g;
+                            btemp[ti * TS + tj] = b;
                         }
-
-                        setUnlessOOG(rtemp[ti * TS + tj], r);
-                        setUnlessOOG(gtemp[ti * TS + tj], g);
-                        setUnlessOOG(btemp[ti * TS + tj], b);
                     }
                 }
```

For the shadows, I'm still testing my current solution -- I'll commit (both) once I'm happy

**Comment by heckflosse** [comment]
@agriggio 

> if (max(r, g, b) > MAXVALF && min(r, g, b) < MAXVALF)

somehow looks familiar

**Comment by agriggio** [comment]
meaning?

**Comment by heckflosse** [comment]
@agriggio Alberto, that was just a thought, as I remembered that I used the same pattern shortly ago for a different part of rt code...

**Comment by agriggio** [comment]
ah, ok... I thought I made some silly mistake :-)

**Comment by heckflosse** [comment]
@agriggio Alberto, it was this, which is still wip

https://github.com/Beep6581/RawTherapee/pull/4444/commits/e03646fc5b0b8746eed1540ddb5137691565dba1#diff-e653e67851b47e016aa6492071be7b5fR2429

**Comment by TooWaBoo** [comment]
@agriggio 
It's different now but still not fixed.

5.4 (Nikon D90)
![2](https://user-images.githubusercontent.com/13800920/38259373-20d3cc22-3765-11e8-8fc4-48de6fab394d.png)

dev (5.4-133-g16679296) (Nikon D90)
![1](https://user-images.githubusercontent.com/13800920/38259437-51a7452c-3765-11e8-8b28-ef205fd0f8ae.png)

5.4 (Nikon P330)
![4](https://user-images.githubusercontent.com/13800920/38259457-5b78447a-3765-11e8-908b-507d890f0e7f.png)

dev (5.4-133-g16679296) (Nikon P330)
![3](https://user-images.githubusercontent.com/13800920/38259497-764687b2-3765-11e8-98b0-edb198f695d3.png)

I'm sorry but I'm not allowed to upload the RAW files.

**Comment by Beep6581** [comment]
`5.4-133-g`16679296 looks fine here:

Neutral:
![screenshot_20180403_175407](https://user-images.githubusercontent.com/5844619/38260531-229d2136-3768-11e8-8fe0-884afe138df0.png)

Neutral with color propagation:
![screenshot_20180403_175420](https://user-images.githubusercontent.com/5844619/38260534-2698fd50-3768-11e8-8d7b-03e8129e3d20.png)

**Comment by agriggio** [comment]
@TooWaBoo can you send them in a private message maybe? (on pixls)

**Comment by TooWaBoo** [comment]
@agriggio 
Message sent on pixls

**Comment by agriggio** [comment]
@TooWaBoo I just pushed a fix -- hopefully this time I covered all cases :-)

**Comment by TooWaBoo** [comment]
@agriggio 
The posterize effect has gone but the "Highlight reconstruction" in RT5.4 is better. I prefere a checkbox old/new behaviour. ;-)
https://filebin.net/oe21qgngapz49ytk

5.4
![5 4](https://user-images.githubusercontent.com/13800920/38321356-b3195f16-3837-11e8-8de2-fbc1fbcb6ae1.png)

dev
![dev](https://user-images.githubusercontent.com/13800920/38321362-bba0e7b2-3837-11e8-974b-3791e041b93d.png)

**Comment by agriggio** [comment]
Just be a bit more generous with highlight compression

**Comment by TooWaBoo** [comment]
Another example with "Color propagation"

Highlight compression = 0
![1](https://user-images.githubusercontent.com/13800920/38340556-61929814-3874-11e8-81ee-38a017f22e50.png)

dev Highlight compression = 150
![3](https://user-images.githubusercontent.com/13800920/38340584-801d44d2-3874-11e8-86b1-119f9770c4e3.png)

5.4 Highlight compression = 150
![2](https://user-images.githubusercontent.com/13800920/38340602-9169e0f6-3874-11e8-9488-6b8ee08279eb.png)

**Comment by Floessie** [comment]
I must admit, I had to go back to 5.4 for a job recently due to the problems with current HLR. So I'm very pleased that @TooWaBoo keeps on bothering and Alberto keeps on optimizing. Whether the whole thing is worth it, I don't know, as the benefits discussed on pixls weren't immediately obvious to me. But I'm no expert here, just a user...

@agriggio @TooWaBoo Please continue to struggle for a solution, that fits best for the casual user and the math-inclined expert. :+1: 

JM2¢,
Flössie

**Comment by agriggio** [comment]
I'm working on a solution

**Comment by agriggio** [comment]
Here you go:

![screenshot from 2018-04-05 14 00 14](https://user-images.githubusercontent.com/11982086/38364676-e3418b62-38d9-11e8-8161-a0482e5d4c38.jpeg)

I hope this makes everyone happy :-)
(this is already in `dev`, btw)

**Comment by Beep6581** [comment]
I like to hover the mouse over a widget and click it to see what changes, which could take a few seconds, then click again to change back, eyes on the preview the whole time. Large tooltips get in the way, obscure and distract. Issue #2221, which nobody responded to, is about moving them away from active widget elements and moving them onto labels where you wouldn't normally statically hover the cursor, and about moving unnecessary tooltips into RawPedia where they are easier to maintain, translate and demonstrate with images. Furthermore, I believe we should use tooltips only for information which is needed to be seen many times, not information which one needs to read only once. For example, hovering over a history panel item shows a tooltip with the whole history message, which is useful since the history panel is narrow and often a row will be truncated as it doesn't fit within the panel. On the other hand, hovering over "Auto Levels" tells you nothing you need to see more than once, and it's a large and intrusive tooltip.

@agriggio would you mind removing that tooltip and putting the info into RawPedia?

**Comment by agriggio** [comment]
@Beep6581 sure, no problem. I wasn't even aware that #2221 existed, otherwise I wouldn't have added the tooltip in the first place...

**Comment by TooWaBoo** [comment]
@agriggio 

Thanks for adding the checkbox. I'm happy now.
... but the checkbox has no effect. It doesn't switch between the two modes.

```
Version: 5.4-144-g7328fa92
Branch: dev
Commit: 7328fa92
Commit date: 2018-04-05
Compiler: gcc 7.3.0
Processor: undefined
System: Windows
Bit depth: 64 bits
Gtkmm: V3.22.0
Lensfun: V0.3.2.0
Build type: release
Build flags: -mwin32 -m64 -mthreads -msse2 -std=c++11 -march=native -Werror=unused-label -fopenmp -Werror=unknown-pragmas -Wall -Wno-unused-result -Wno-deprecated-declarations -Wno-aggressive-loop-optimizations -DNDEBUG -O3 -ftree-vectorize
Link flags: -m64 -mthreads -static-libgcc -march=native -s -O3
OpenMP support: ON
MMAP support: ON
```
